### DISCUSSION TITLE
Test cases/setups

### DISCUSSION THREAD

**Discussion by captnCC** [kube-hetzner/terraform-hcloud-kube-hetzner#405] [Ideas]
In an attempt do ease the process to test and verify on merge requests it might be good idea to start with a list of test setups that could be run at least locally against own hetzner accounts. I remember there was an PR to provide testing with go, but I can't find it anymore.

So I will just start with the cases I can imagine and which have causes problems in the past:

| Controlplanes | Agents | Autoscaling | Addons | Notes |
|---|---|---|---|---|
| 1 | 0 | none | none| single node cluster |
| 1 | 1 | none | none| non-ha setup | 
| 1 | 0 | enabled | none| pure autoscaling agents |
| 3 | 1 | none | none | HA-controlplanes |
| 3 | 1 | enabled | none | HA-controlplanes + mixed static & scaling agents |

The list could continue with a variation of addons, but I don't have any expertise in that. Especially the storage options with longhorn and/or hetzner volumes would make a good addition to create more trust in their functionality!

**Comment by phaer** [comment]
> I remember there was an PR to provide testing with go, but I can't find it anymore.

https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/pull/153

**Comment by captnCC** [comment]
Maybe another general question on the topic how to test: Is a compiled binary like go the way to go or would something in the realm of bash be a better fit. 
As my opinion would be in favour of bash for the following reasons:
- Simplicity in the toolchain: bash just runs everywhere without setup
- No additional knowledge needed (most people working in the realm of [dev]ops will be familiar with it)

The drawback would be: bash .. it can get messy real quick

**Reply by phaer** [reply]
go compilation is quick enough that it doesn't make a noticeable difference in practice and pretty much all components we'd like to test are written in go as well. Python might be viable alternative, but meaningful tests in bash won't be much fun. 
Just try to to (re-)implement it if wanted ;)

**Reply by captnCC** [reply]
The compilation time wasn't my concern, more over the toolchain needed to compile go. Maybe a docker-dev setup that would allow a "zero-requirement" usage could be helpful. (I have basically zero knowledge of go 🙈)

**Reply by phaer** [reply]
The "toolchain" you'd need for go is a single binary, anything else (like the tests build-time dependencies) is automatically  fetched during runtime. Docker alone is more complex than that

**Reply by captnCC** [reply]
I think its time for me to check it finally out!

**Reply by mysticaltech** [reply]
@captnCC Thanks for re-viving the matter. Your proposal of testing in one's own account is a great idea! 

I had kept the `add-testsuite` branch just for such a use case, as @phaer did great work there. (but someone else on the team had deleted it). Now just restaured it, you can find it here https://github.com/kube-hetzner/terraform-hcloud-kube-hetzner/tree/add-testsuite.

About the choice of language, whatever you see fit yourself. I myself am more of a Python dev, but Go is probably worth learning when you work with Kubernetes.

**Comment by mysticaltech** [comment]
@captnCC This would really help all of us -- You do not know how much of my time goes into testing all possible use cases for each major PR, it's just a pain in the bum! 🤦 So really glad you are reviving this... 🙏 

(please do not miss the add-testsuite banch, if you wish to build upon phaer's work, or at least inspire yourself from it). Both Python and Go seems decent for me. Bash on the other hand, I think it's too low level.

**Reply by phaer** [reply]
Just another note for people working on testing kube-hetzner: https://github.com/syself/cluster-api-provider-hetzner mentions that hetzner sponsors e2e test infrastructure for them. Might be worth asking them ;)

(won't do it myself as I am not using kube-hetzner myself anymore)

**Reply by mysticaltech** [reply]
Good to know! I will ask, as soon as we have something working!

**Reply by ifeulner** [reply]
> (won't do it myself as I am not using kube-hetzner myself anymore)

@phaer Why is this? What you are using now?

**Reply by phaer** [reply]
@ifeulner I have no need for kubernetes anymore, just using normal nixos boxes.

**Reply by mysticaltech** [reply]
@phaer But it's still nice that we still get to pick your brain on important issues! Thanks for that... 🙏💙
